test/metrics/metrics.jl

```diff
@@ -0,0 +1,215 @@
+
+if !@isdefined(TEST_RS)
+    const TEST_DOM, TEST_N_SAMPLES, TEST_SCENS, TEST_RS = test_rs()
+end
+
+function _test_metrics(
+    metric::metrics.Metric, params::Tuple; relative::Bool=false
+)::Nothing
+    metric_result = metric(params...)
+    @test all(haskey.([metric_result.properties], (:metric_name, :metric_unit)))
+    @test all(0.0 .<= metric_result.data)
+    relative && @test all(metric_result.data .<= 1.0)
+    return nothing
+end
+
+function test_relative_metric(metric::metrics.Metric, params::Tuple)::Nothing
+    return _test_metrics(metric, params; relative=true)
+end
+
+function test_absolute_metric(metric::metrics.Metric, params::Tuple)::Nothing
+    return _test_metrics(metric, params; relative=false)
+end
+
+@testset "metrics.jl" begin
+    n_timesteps::Int64 = length(TEST_RS.env_layer_md.timeframe)
+    n_groups::Int64 = length(ADRIA.coral_spec().taxa_names)
+    n_sizes::Int64 = length(ADRIA.coral_spec().params.name) / n_groups
+    n_group_sizes::Int64 = n_groups * n_sizes
+    n_locations::Int64 = length(TEST_RS.coral_dhw_tol_log.locations)
+    n_scenarios::Int64 = size(TEST_SCENS, 1)
+    test_scens_datacube::YAXArray{Float64,2} = DataCube(
+        Matrix(TEST_SCENS); scenarios=1:n_scenarios, factors=names(TEST_SCENS)
+    )
+    cover_params = (
+        n_timesteps=n_timesteps,
+        n_group_sizes=n_group_sizes,
+        n_locations=n_locations,
+        n_scenarios=n_scenarios
+    )
+
+    coral_cover::YAXArray{Float64,4} = Factories.coral_cover(; cover_params...)
+    zero_coral_cover::YAXArray{Float64,4} = Factories.zero_coral_cover(; cover_params...)
+    full_coral_cover::YAXArray{Float64,4} = Factories.full_coral_cover(; cover_params...)
+    test_covers::Vector{YAXArray{Float64,4}} = [
+        coral_cover, zero_coral_cover, full_coral_cover
+    ]
+    k_area::Vector{Float64} = round.(rand(n_locations) .* 100; digits=2)
+
+    @testset "relative_cover" begin
+        test_relative_metric(metrics.relative_cover, (TEST_RS,))
+        for cover in test_covers
+            test_relative_metric(metrics.relative_cover, (cover,))
+        end
+    end
+
+    @testset "total_absolute_cover" begin
+        test_absolute_metric(metrics.total_absolute_cover, (TEST_RS,))
+        for cover in test_covers
+            test_absolute_metric(
+                metrics.total_absolute_cover, (metrics.relative_cover(cover), k_area)
+            )
+        end
+    end
+
+    @testset "relative_taxa_cover" begin
+        test_relative_metric(metrics.relative_taxa_cover, (TEST_RS,))
+        for cover in test_covers
+            test_relative_metric(
+                metrics.relative_taxa_cover, (cover[:, :, :, 1], k_area, n_groups)
+            )
+        end
+    end
+
+    @testset "relative_loc_taxa_cover" begin
+        for cover in test_covers
+            test_relative_metric(
+                metrics.relative_loc_taxa_cover, (cover[:, :, :, 1], k_area, n_groups)
+            )
+        end
+    end
```

Comment:
For these ones that don't need any special setup I wonder if you can simply generate the tests leveraging the metadata?

```julia
using ADRIA

all_metrics = names(metrics, all=true)
metric_identifiers = ADRIA.human_readable_name(all_metrics)

relative_metrics = occursin.("relative", metric_identifiers)
absolute_metrics = occursin.("absolute", metric_identifiers)
test_m_id = vcat(findall.([relative_metrics, absolute_metrics])...)

for m_id in test_m_id
    metric_to_test = getfield(metrics, all_metrics[m_id])
    params = ...  # infer appropriate mock data from metric metadata?

    # This would apply the appropriate test if relative or not
    _test_metrics(metric_to_test, params)

    # Or keep the flag
    _test_metrics(metric_to_test, params; is_relative=relative_metrics[m_id])
end


```

Reply:
One issue I have with the current approach is that you have to manually maintain all the tests, which will be a pain.

The "vision" for the metrics were to define a "base" set of metrics and the rest would be auto-generated (including tests). I just never got around to sitting down and working out the necessary meta-programming to do this.

Reply:
few comments about this approach: 
1. there are some metrics that I consider to be "absolute" that don't have the word "absolute" in the name, like `juvenile_indicator` or `coral_evenness`. But we could manually add those to the vector of metrics.
2. The setup for each metric is slightly different. For the functions that accept just a ResultSet as input, that's ok. But for the ones that I need to pass a list of arguments, I would actually need to do this setup one by one. 
3. These are just an initial set of tests, but we might want to write more specific tests for each metric (like the `relative_shelter_volume` that has a test scenario only with coral with the maximum shelter density).

Reply:
> About item 2: I see in your code snippet you wrote params = ...  # infer appropriate mock data from metric metadata?. I guess my question to that is: how?

We can give as much detail in the metadata we need to make this all work seamlessly. Marking each metric as to absolute or relative nature could also address the need to infer it from their name alone, which is a more robust approach too...

I'm throwing out ideas, that's all.

Reply:
I should clarify that the auto-generation of tests doesn't necessarily need to be in this PR.